crates/goose/src/agents/default.rs

```diff
@@ -63,77 +61,47 @@ impl DefaultAgent {
         if approx_count > target_limit {
             println!("[WARNING] Token budget exceeded. Current count: {} \n Difference: {} tokens over buget. Removing context", approx_count, approx_count - target_limit);
 
-            // Get token counts for each resource
-            let mut system_token_counts = HashMap::new();
-
-            // Iterate through each system and its resources
-            for (system_name, resources) in resource_content {
-                let mut resource_counts = HashMap::new();
-                for (uri, (_resource, content)) in resources {
-                    let token_count =
-                        self.token_counter.count_tokens(content, Some(model_name)) as u32;
-                    resource_counts.insert(uri.clone(), token_count);
-                }
-                system_token_counts.insert(system_name.clone(), resource_counts);
-            }
-
-            // Sort resources by priority and timestamp and trim to fit context limit
-            let mut all_resources: Vec<(String, String, Resource, u32)> = Vec::new();
-            for (system_name, resources) in resource_content {
-                for (uri, (resource, _)) in resources {
-                    if let Some(token_count) = system_token_counts
-                        .get(system_name)
-                        .and_then(|counts| counts.get(uri))
-                    {
-                        all_resources.push((
-                            system_name.clone(),
-                            uri.clone(),
-                            resource.clone(),
-                            *token_count,
-                        ));
-                    }
+            for item in resource_items.iter_mut() {
+                if item.token_count.is_none() {
+                    let count = self
+                        .token_counter
+                        .count_tokens(&item.content, Some(model_name))
+                        as u32;
+                    item.token_count = Some(count);
                 }
             }
 
             // Sort by priority (high to low) and timestamp (newest to oldest)
-            all_resources.sort_by(|a, b| {
-                let a_priority = a.2.priority().unwrap_or(0.0);
-                let b_priority = b.2.priority().unwrap_or(0.0);
-                if (b_priority - a_priority).abs() < PRIORITY_EPSILON {
-                    b.2.timestamp().cmp(&a.2.timestamp())
+            let mut all_items: Vec<ResourceItem> = resource_items.to_vec();
+            all_items.sort_by(|a, b| {
+                // Compare priority first (descending)
+                let diff = b.priority - a.priority;
```

Comment:
nit: no point to sort by priority as well if we only include active resources (priority=1)?